### ISSUE TITLE
"module yt_dlp.dependencies.Cryptodome has no attribute Cipher" error after upgrading from 2023.01.06 to 2023.02.17 via pip

### ISSUE THREAD

**Issue by garettharnish** [yt-dlp/yt-dlp#6272]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2023.02.17** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

I'm having a dependency issue on the newest version of yt_dlp (see Verbose Output).  I tried purging the pip3 cache and uninstalling yt_dlp and it's dependencies then reinstalling:

`python3 -m pip install yt_dlp
Collecting yt_dlp
  Downloading yt_dlp-2023.2.17-py2.py3-none-any.whl (2.9 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.9/2.9 MB 10.2 MB/s eta 0:00:00
Collecting pycryptodomex
  Downloading pycryptodomex-3.17-cp35-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (2.1 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 32.2 MB/s eta 0:00:00
Collecting websockets
  Downloading websockets-10.4-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (106 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 106.7/106.7 kB 11.7 MB/s eta 0:00:00
Collecting brotli
  Downloading Brotli-1.0.9-cp37-cp37m-manylinux1_x86_64.whl (357 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 357.2/357.2 kB 24.9 MB/s eta 0:00:00
Collecting mutagen
  Downloading mutagen-1.46.0-py3-none-any.whl (193 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.6/193.6 kB 19.2 MB/s eta 0:00:00
Collecting certifi
  Downloading certifi-2022.12.7-py3-none-any.whl (155 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 155.3/155.3 kB 17.2 MB/s eta 0:00:00
Installing collected packages: brotli, websockets, pycryptodomex, mutagen, certifi, yt_dlp
Successfully installed brotli-1.0.9 certifi-2022.12.7 mutagen-1.46.0 pycryptodomex-3.17 websockets-10.4 yt_dlp-2023.2.17
`

Error remains.  Reverting to yt_dlp 2023.01.06 makes error go away.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/yt_dlp/compat/compat_utils.py", line 36, in __getattribute__
    ret = super().__getattribute__(attr)
  File "/usr/local/lib/python3.7/site-packages/yt_dlp/compat/compat_utils.py", line 56, in __getattr__
    raise AttributeError(f'module {parent.__name__} has no attribute {attr}')
AttributeError: module yt_dlp.dependencies.Cryptodome has no attribute Cipher

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/yt-dlp", line 5, in <module>
    from yt_dlp import main
  File "/usr/local/lib/python3.7/site-packages/yt_dlp/__init__.py", line 18, in <module>
    from .cookies import SUPPORTED_BROWSERS, SUPPORTED_KEYRINGS
  File "/usr/local/lib/python3.7/site-packages/yt_dlp/cookies.py", line 18, in <module>
    from .aes import (
  File "/usr/local/lib/python3.7/site-packages/yt_dlp/aes.py", line 5, in <module>
    from .dependencies import Cryptodome
  File "/usr/local/lib/python3.7/site-packages/yt_dlp/dependencies/__init__.py", line 76, in <module>
    Cryptodome_AES = Cryptodome.Cipher.AES if Cryptodome else None
  File "/usr/local/lib/python3.7/site-packages/yt_dlp/compat/compat_utils.py", line 43, in __getattribute__
    ret = getter(attr)
  File "/usr/local/lib/python3.7/site-packages/yt_dlp/compat/compat_utils.py", line 56, in __getattr__
    raise AttributeError(f'module {parent.__name__} has no attribute {attr}')
AttributeError: module yt_dlp.dependencies.Cryptodome has no attribute Cipher
```

**Comment by pukkandan** [comment]
Does `pip install cffi` fix the issue?

**Comment by pukkandan** [comment]
> Error remains. Reverting to yt_dlp 2023.01.06 makes error go away.

Also, could you give verbose log of this?

**Comment by And4713** [comment]
I am also getting this error message and I don't even use pip. Python 3.8, `AttributeError: module yt_dlp.dependencies.Cryptodome.Cipher has no attribute AES`

**Comment by YummyOreo** [comment]
I am also getting this issue, but only when using `pyinstaller`. Installing `cffi` does not do anything.

**Comment by pukkandan** [comment]
"Me too"s without additional details isn't helping anyone. I have yet to be able to reproduce the issue.

If you are having the issue, run the following and post their outputs:
* `yt-dlp -v` on both the new and old versions (use `pip install yt-dlp==2023.1.6` to install old version)
* If on PIP (change `py` to `python`/`python3` if needed):
	* `pip install cffi`
    * `py -c "from Cryptodome.Cipher import AES; print(AES)"`
    * `py -c "from Crypto.Cipher import AES; print(AES)"`
    * (`pip freeze` may also help)
* If compiled yourself (`pyinstaller` etc)
	* Everything from PIP section
	* Post exact steps of how you compiled it

My guess is that cryptodome install is somehow broken in your systems and previous versions of yt-dlp were silently ignoring the error[^1]. If I am right, it's an easy fix to make yt-dlp ignore the error again. But first I need to be sure

[^1]: There was a `try: except ImportError` around entire import call

**Comment by garettharnish** [comment]
I stayed downgraded at 2023.01.06 after filing the report.

yt-dlp -v
[debug] Command-line config: ['-v']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out UTF-8, error UTF-8, screen UTF-8
[debug] yt-dlp version 2023.01.06 [6becd25] (pip)
[debug] Python 3.7.15 (CPython x86_64 64bit) - Linux-3.10.0-1160.15.2.el7.x86_64-x86_64-with-centos-7.9.2009-Core (OpenSSL 1.0.2k-fips  26 Jan 2017, glibc 2.17)
[debug] exe versions: ffmpeg 5.0.1 (fdk,setts), ffprobe 5.0.1
[debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[debug] Proxy map: {}
[debug] Loaded 1760 extractors

I then decided to skip and do the install of cffi and run the commands for you:


$ python3 -c "from Cryptodome.Cipher import AES; print(AES)"
<module 'Cryptodome.Cipher.AES' from '/usr/local/lib/python3.7/site-packages/Cryptodome/Cipher/AES.py'>

$ python3 -c "from Crypto.Cipher import AES; print(AES)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'Crypto'

I then upgraded again to 2023.02.17 and it isn't crashing:

$ yt-dlp -v
[debug] Command-line config: ['-v']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out UTF-8, error UTF-8, screen UTF-8
[debug] yt-dlp version 2023.02.17 [a0a7c01] (pip)
[debug] Python 3.7.15 (CPython x86_64 64bit) - Linux-3.10.0-1160.15.2.el7.x86_64-x86_64-with-centos-7.9.2009-Core (OpenSSL 1.0.2k-fips  26 Jan 2017, glibc 2.17)
[debug] exe versions: ffmpeg 5.0.1 (fdk,setts), ffprobe 5.0.1
[debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[debug] Proxy map: {}
[debug] Loaded 1782 extractors

Uninstalling cffi causes it to crash again.   With cffi uninstalled:

$ python3 -c "from Cryptodome.Cipher import AES; print(AES)"
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/Cryptodome/Util/_raw_api.py", line 79, in <module>
    from cffi import FFI
ModuleNotFoundError: No module named 'cffi'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.7/site-packages/Cryptodome/Cipher/__init__.py", line 27, in <module>
    from Cryptodome.Cipher._mode_ecb import _create_ecb_cipher
  File "/usr/local/lib/python3.7/site-packages/Cryptodome/Cipher/_mode_ecb.py", line 29, in <module>
    from Cryptodome.Util._raw_api import (load_pycryptodome_raw_lib,
  File "/usr/local/lib/python3.7/site-packages/Cryptodome/Util/_raw_api.py", line 164, in <module>
    import ctypes
  File "/usr/local/lib/python3.7/ctypes/__init__.py", line 7, in <module>
    from _ctypes import Union, Structure, Array
ModuleNotFoundError: No module named '_ctypes'

So yeah, Cryptodome was broken.  Installing cffi resolved the issue.

**Comment by garettharnish** [comment]
So I consider this resolved.  Strange the forcing it to reinstall pycryptodome didn't install cffi.

**Comment by Ghrabo** [comment]
@pukkandan Could you please share this easy solution that you mentioned: "If I am right, it's an easy fix to make yt-dlp ignore the error again."?
I've been trying to fix the problem the proper way as you guys described but it is not working. I am getting errors even installing cffi. If there is an easier workaround could you please share it? I just want to have yt-dlp working again.

**Comment by pukkandan** [comment]
Yet, you refuse to give the information I asked for?

See the diff in my above comment

**Comment by egzn** [comment]
Can confirm that @pukkandan's [solution](https://github.com/yt-dlp/yt-dlp/issues/6272#issuecomment-1435433009) works.

**Comment by And4713** [comment]
In case it helps anyone else what ended up fixing the problem for me was `python3.8 -m pip install cryptodomex`
I am using deadsnakes alternate python versions. Given this thread indicated the installation was likely the problem led to that solution.

**Comment by Ghrabo** [comment]
My issue is resolved as well now. As I was starting to gather the information that @pukkandan asked for, it looks like the following two commands completely fixed my problem (use pip to downgrade yt-dlp to 2023.1.6, and then upgrade it back to 2023.02.17):
`pip install yt-dlp==2023.1.6`
`pip install yt-dlp==2023.02.17`

This seemed to properly install yt-dlp together will all needed dependencies, including pycryptodomex and mutagen.

The way I was using yt-dlp previously was not through pip, but from the release binaries. And then I was running yt-dlp -U to keep upgrading it to the latest version. I guess at some point something went wrong and then this issue showed up. But now, reinstalling yt-dlp through pip has resolved the issue.

In the end I did not have to resort to using @pukkandan's diff solution workaround.

**Comment by pukkandan** [comment]
https://github.com/yt-dlp/yt-dlp/issues/6185#issuecomment-1438384918

**Comment by psette** [comment]
I tried downgrading and then upgrading, doing the fix [here](https://github.com/yt-dlp/yt-dlp/issues/6272#issuecomment-1435433009). this runs fine locally, but when i deploy it to AWS lambda now it fails with the following error. Any suggestions?


`[ERROR] OSError: Cannot load native module 'Cryptodome.Cipher._raw_ecb': Not found '_raw_ecb.cpython-39-x86_64-linux-gnu.so',`

data:

>  (env) ➜  lambdas git:(master) ✗ yt-dlp -v
> [debug] Command-line config: ['-v']
> [debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
> [debug] yt-dlp version 2023.02.17 [a0a7c01] (pip)
> [debug] Python 3.8.9 (CPython x86_64 64bit) - macOS-12.1-x86_64-i386-64bit (LibreSSL 2.8.3)
> [debug] exe versions: ffmpeg 5.1.2 (setts), ffprobe 5.1.2
> [debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.1
> [debug] Proxy map: {}
> [debug] Loaded 1782 extractors
> 
> Usage: yt-dlp [OPTIONS] URL [URL...]
> 
> yt-dlp: error: You must provide at least one URL.
> Type yt-dlp --help to see a list of all options.
> 

<br class="Apple-interchange-newline">

> (env) ➜  lambdas git:(master) ✗ python3  -c "from Cryptodome.Cipher import AES; print(AES)"
<module 'Cryptodome.Cipher.AES' from '/Users/pietrosette/Documents/autoclip/lambdas/env/lib/python3.8/site-packages/Cryptodome/Cipher/AES.py'>

> (env) ➜  lambdas git:(master) ✗ python3 -c "from Crypto.Cipher import AES; print(AES)"
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
> ModuleNotFoundError: No module named 'Crypto'

**Comment by pukkandan** [comment]
You are on an older version

**Comment by psette** [comment]
@pukkandan, I tried updating the versions, but unfortunately, that didn't fix the problem. What ended up working was building the code on an EC2 instance using CodePipeline.

I have an M1 Mac, and when I built the code locally and then deployed it to the Lambda function, it broke (even though I was using the same versions and code). However, building the code through CodePipeline solved the issue. I suspect that there might be something with how the package is reading the reading the architecture to create the binary that is causing the problem.
